File: module-8-harness/solutions/session.ts
/**
 * Module 8: Session persistence
 * Save/load/list sessions as JSON for crash recovery.
 */
import * as fs from "fs/promises";
import * as path from "path";
import type { Message } from "../../module-2-agent-loop/solutions/types.js";

export interface Session {
  id: string;
  createdAt: string;
  updatedAt: string;
  task: string;
  messages: Message[];
  turn: number;
  workspace: string;
}

const SESSIONS_DIR = path.join(process.cwd(), "sessions");

async function ensureDir() {
  await fs.mkdir(SESSIONS_DIR, { recursive: true });
}

export async function saveSession(session: Session): Promise<void> {
  await ensureDir();
  session.updatedAt = new Date().toISOString();
  await fs.writeFile(
    path.join(SESSIONS_DIR, `${session.id}.json`),
    JSON.stringify(session, null, 2),
    "utf-8"
  );
}

export async function loadSession(id: string): Promise<Session | null> {
  try {
    const data = await fs.readFile(path.join(SESSIONS_DIR, `${id}.json`), "utf-8");
    return JSON.parse(data);
  } catch {
    return null;
  }
}

export async function listSessions(): Promise<{ id: string; task: string; updatedAt: string }[]> {
  await ensureDir();
  const files = await fs.readdir(SESSIONS_DIR);
  const sessions = [];
  for (const file of files) {
    if (!file.endsWith(".json")) continue;
    try {
      const data = await fs.readFile(path.join(SESSIONS_DIR, file), "utf-8");
      const s = JSON.parse(data);
      sessions.push({ id: s.id, task: s.task, updatedAt: s.updatedAt });
    } catch {
      // skip corrupt files
    }
  }
  return sessions.sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
}

export function createSession(task: string, workspace: string): Session {
  return {
    id: `session-${Date.now()}`,
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString(),
    task,
    messages: [{ role: "user", content: task }],
    turn: 0,
    workspace,
  };
}

File: module-8-harness/solutions/python/session.py
"""Module 8: Session persistence."""
import os
import json
import time

SESSIONS_DIR = os.path.join(os.getcwd(), "sessions")


def _ensure_dir():
    os.makedirs(SESSIONS_DIR, exist_ok=True)


def save_session(session: dict) -> None:
    _ensure_dir()
    from datetime import datetime
    session["updatedAt"] = datetime.now().isoformat()
    path = os.path.join(SESSIONS_DIR, f"{session['id']}.json")
    with open(path, "w") as f:
        json.dump(session, f, indent=2, default=str)


def load_session(session_id: str) -> dict | None:
    try:
        with open(os.path.join(SESSIONS_DIR, f"{session_id}.json")) as f:
            return json.load(f)
    except FileNotFoundError:
        return None


def list_sessions() -> list[dict]:
    _ensure_dir()
    sessions = []
    for fname in os.listdir(SESSIONS_DIR):
        if not fname.endswith(".json"):
            continue
        try:
            with open(os.path.join(SESSIONS_DIR, fname)) as f:
                s = json.load(f)
            sessions.append({"id": s["id"], "task": s["task"], "updatedAt": s["updatedAt"]})
        except Exception:
            pass
    return sorted(sessions, key=lambda s: s["updatedAt"], reverse=True)


def create_session(task: str, workspace: str) -> dict:
    from datetime import datetime
    now = datetime.now().isoformat()
    return {
        "id": f"session-{int(time.time() * 1000)}",
        "createdAt": now,
        "updatedAt": now,
        "task": task,
        "messages": [{"role": "user", "content": task}],
        "turn": 0,
        "workspace": workspace,
    }
